website/photos/services.py

```diff
@@ -70,3 +84,101 @@ def get_annotated_accessible_albums(request, albums):
         )
 
     return albums
+
+
+def extract_archive(request, album, archive):
+    iszipfile = is_zipfile(archive)
+    archive.seek(0)
+
+    if iszipfile:
+        with ZipFile(archive) as zip_file:
+            for photo in sorted(zip_file.infolist(), key=lambda x: x.filename):
+                extract_photo(request, zip_file, photo, album)
+    else:
+        # is_tarfile only supports filenames, so we cannot use that
+        try:
+            with tarfile.open(fileobj=archive) as tar_file:
+                for photo in sorted(tar_file.getmembers(), key=lambda x: x.name):
+                    extract_photo(request, tar_file, photo, album)
+        except tarfile.ReadError:
+            raise ValueError(_("The uploaded file is not a zip or tar file."))
+
+
+def extract_photo(request, archive_file, photo, album):
+    # zipfile and tarfile are inconsistent
+    if isinstance(photo, ZipInfo):
+        photo_filename = photo.filename
+        extract_file = archive_file.open
+    elif isinstance(photo, tarfile.TarInfo):
+        photo_filename = photo.name
+        extract_file = archive_file.extractfile
+    else:
+        raise TypeError("'photo' must be a ZipInfo or TarInfo object.")
+
+    # Ignore directories
+    if not os.path.basename(photo_filename):
+        return
+
+    # Generate unique filename
+    num = album.photo_set.count()
+    filename, extension = os.path.splitext(photo_filename)
+    new_filename = str(num).zfill(4) + extension
+
+    photo_obj = Photo()
+    photo_obj.album = album
+    try:
+        with extract_file(photo) as f:
+            photo_obj.file.save(new_filename, File(f))
+
+        if not save_photo(photo_obj):
+            messages.add_message(
+                request, messages.WARNING, _("{} is duplicate.").format(photo_filename)
+            )
+    except (OSError, AttributeError, UnidentifiedImageError):
+        messages.add_message(
+            request, messages.WARNING, _("Ignoring {}").format(photo_filename)
+        )
+        if photo_obj.file.path:
+            os.remove(photo_obj.file.path)
+        photo_obj.delete()
+
+
+def save_photo(photo_obj):
+    if photo_obj.original_file != photo_obj.file.path:
```

Comment:
`original_file` is only defined on row 177, however, I can't find any reason for this check being there. This would stop the file from being called two times on the same photo_obj in the same request. However, that can't happen as it is only called in `extract_photo` or in `save_model`, both of which use a new Photo object every time they call `save_photo`. Thus, why is this check here?

Reply:
I think because it was here: https://github.com/svthalia/concrexit/pull/1059/files/c80d975b0d0208971969a054f5252ea2d850424f#diff-55499884164694c3509b150ab9e9b6a6L65

Reply:
The check exists so that we only try to update the photos that are actually changed, you can save the photo without changing the file and this prevents it from saving. However, I do think that think check does not belong here since we are already saving the photo, and it is only relevant when updating a Photo object. So it should probably move to the save_model of the PhotoAdmin

Reply:
Moved it to the right place.